### ISSUE TITLE
consider removing unused bootstrap styles

### ISSUE THREAD

**Issue by leeoniya** [krausest/js-framework-benchmark#100]
a large part of this benchmark is browser restyling overhead which has no bearing on lib speed and is a fixed unnecessary cost for everyone. in many cases it also masks lib efficiency. if for no other reason than to allow the benchmark runs to complete in less time, it may make sense to strip out unused styles.

you can use something like https://github.com/purifycss/purifycss to avoid having to figure things out by hand.

**Comment by leeoniya** [comment]
i can work on this and mame a PR if you're interested

**Comment by krausest** [comment]
Would be great!

**Comment by leeoniya** [comment]
i'm at 10.8k non-minified vs the original bootstrap 118k *minified*.

there's still a bunch that can be removed, a bunch of it is `@media` responsive styles plus some `@keyframes` stuff.

should i strip out the `@media` stuff for small screens?

**Comment by krausest** [comment]
I'd say so. Thanks!

**Comment by leeoniya** [comment]
before i do any manual pruning can you see if you get any improvement from the attached versions? i'm not seeing too much of a boost.

drop them into `/css/bootstrap/dist/css`

[js-framework-benchmark.css.zip](https://github.com/krausest/js-framework-benchmark/files/640763/js-framework-benchmark.css.zip)

**Comment by krausest** [comment]
Hard to say whether it's really faster. Ran it a few times and the original css is often about 2% slower, but it doubt whether is significant. 
<img width="240" alt="bildschirmfoto 2016-12-09 um 21 06 22" src="https://cloud.githubusercontent.com/assets/3374055/21062949/7d24008c-be53-11e6-87fd-e1b9450d892d.png">

**Comment by leeoniya** [comment]
what about with no styling at all? <- this should really be the baseline.

**Comment by krausest** [comment]
Ok. I needed to add the following rules to main.css and left out bootstrap
```
.danger {
    background-color: #faa;
}

.glyphicon-remove:before {
    content: "x";
}
```
and then got the following results:
<img width="299" alt="bildschirmfoto 2016-12-09 um 22 01 46" src="https://cloud.githubusercontent.com/assets/3374055/21064472/3bd71422-be5b-11e6-83fa-e75a6508e58f.png">

**Comment by leeoniya** [comment]
we should see if there's some specific rule that causes the huge slowdown. sometimes subtle things like `box-shadow` or `opacity` will cause the painting to go into overdrive. it may not be the selector matching but rather some specific rule.

the best way would probably be to just build the css upwards by hand and track the perf overhead at each step. i suspect there's something really specific that's responsible for the majority of the drop here.

**Comment by krausest** [comment]
Might be due to some missing fixed layout in the table.

**Comment by leeoniya** [comment]
@krausest 

ok, i had some time to dig into it.

the major killers are:

- zebra-striping `tr:nth-child(odd)`
- glyphicons & font metrics
- table-layout: fixed helps a bit

removing zebra striping improves things significantly, especially for row removal where "odd" logically changes for the entire table. not sure if you want to kill it off, but now you know.

the linked `main.css` [1] matches the current style and boosts render time by ~30%. it probably has a few extra button rules that arent needed, feel free to clean it further if you'd like.

[1] http://pastebin.com/919uhYza

**Comment by krausest** [comment]
I tried with your main.css (column is labeled "vanillajs-small-css"):
<img width="350" alt="bildschirmfoto 2016-12-20 um 22 09 34" src="https://cloud.githubusercontent.com/assets/3374055/21368098/32d71d8c-c701-11e6-99b1-fb0cc623f20a.png">
I think I'll take that css when reworking the benchmark.

**Comment by leeoniya** [comment]
@krausest ping!

**Comment by krausest** [comment]
Had that on my todo-list, but you got me ;-)

I'm currently thinking about a script that allows to toggle between bootstrap and the minimal css. This would allow switching without changing the code for the benchmarks. What do you think?

**Comment by leeoniya** [comment]
any reason not to just do a mass search/replace across the repo? why keep the original at all?

**Comment by krausest** [comment]
That was easy (actually not really for ember, polymer and bobril ...)
I really wanted to keep bootstrap and a minimal CSS. You can select which style to use with `node configureStyles.js --minimal | --bootstrap`
I've updated the current `css/useMinimalCss.css` to look more like the bootstrap version - but I'm not sure whether that's a good idea. Might be interesting to have non-alternating row colors and see what strategy for remove row works best then. Results are here for [minimal](https://rawgit.com/krausest/js-framework-benchmark/master/webdriver-ts/table_minimal.html) and [bootstrap](https://rawgit.com/krausest/js-framework-benchmark/master/webdriver-ts/table_bootstrap.html).
I'll have to verify those results still (and find out why kivi is that much faster than vanillajs for the minimal CSS) especially for cases when there's a large difference between the two css files.

**Comment by krausest** [comment]
I thought about switching to minimal css, but after testing a bit I'm not so sure about it and need feedback (ping @localvoid).

It all started with the results posted above. Kivi perfomed create rows in 98 msecs and vanillajs in 116 msecs for minimal CSS and I wanted to see if I can speed up vanillajs. I think I can't as the difference is large due to a large variance between the runs. Thus for this analysis I decided not to drop the 4 slowest runs, but keep all results.

For create rows the minimal CSS can be much faster than bootstrap (say 97 msecs instead of 129 msecs for vanillajs), but at the same time the slow case can be worse too (say 170 msecs for minimal css and vanillajs and 153 msecs for bootstrap and vanilla) and the best case doesn't happen often so the average isn't that good. 
Looking at the results of some frameworks for minimal css and create rows:
 
<img width="529" alt="bildschirmfoto 2017-03-15 um 21 01 24s" src="https://cloud.githubusercontent.com/assets/3374055/23968771/719d6a52-09c4-11e7-863c-40cf71aeb922.png">
And here for bootstrap:
<img width="520" alt="bildschirmfoto 2017-03-15 um 21 04 21b" src="https://cloud.githubusercontent.com/assets/3374055/23968784/827216de-09c4-11e7-83c7-c9faba1ca3ed.png">
Results are really puzzling. Variance is pretty high for (k)ivi, domvm and vanillajs. Inferno is slower for minimal css, but has a much lower variance. The measurement itself seems ok, the timeline view shows similar varying results.

If I force a GC before running the variance goes down, but the fast results are gone too...

With forced GC after init benchmark  the results for minimal CSS look like that:
<img width="536" alt="bildschirmfoto 2017-03-15 um 21 24 20s" src="https://cloud.githubusercontent.com/assets/3374055/23969302/479f7cc0-09c6-11e7-9b8b-a8f4d132651c.png">
And with forced GC for bootstrap:
<img width="532" alt="bildschirmfoto 2017-03-15 um 21 26 14b" src="https://cloud.githubusercontent.com/assets/3374055/23969314/5147423a-09c6-11e7-96b0-fc06ecfec4be.png">
In other words, now minimal CSS is consistently slower than bootstrap. Variance between both is similar. The forced GC hurts minimal CSS, but not bootstrap. This hurts my mind. Any idea what the cause might be?

**Comment by localvoid** [comment]
I have no idea :) Maybe try to look at what is happening in chrome://tracing , start recording a new trace "Javascript and rendering", switch to benchmark tab, reload page/create rows multiple times until you catch this huge difference in results, then stop recording, scroll down to renderer with a benchmark tab, select regions and look where this huge difference comes from.

**Comment by leeoniya** [comment]
@krausest wanna give this another go with the current iteration count and t-test/median measuring in place?

if you have no intention of making minimalcss the default, then it's probably time to close this issue. i dont think there's much use in keeping around 2 css options.

**Comment by krausest** [comment]
I did that - but results are very puzzling
Here's the median for 10 runs with minimal css (and the conclusion is no different for the mean)
<img width="966" alt="minimal_median" src="https://user-images.githubusercontent.com/3374055/28135360-31f923a2-6746-11e7-81ab-d3ae83826331.png">
and here for bootstrap
<img width="987" alt="bootstrap_median" src="https://user-images.githubusercontent.com/3374055/28135370-3a66d246-6746-11e7-996d-a62065c89630.png">
Now with chrome 59 bootstrap appears to be faster than the minimal css (I tried several times, because I couldn't believe)
Unless you convince me I made a big mistake I won't switch to the minimal bootstrap like css.

**Comment by leeoniya** [comment]
> Now with chrome 59 bootstrap appears to be faster than the minimal css (I tried several times, because I couldn't believe)

maybe there's some subtle rule in bootstrap that significantly reduces the restyling load but doesnt have an obvious visual effect (kinda like translate3d to force hw accel or `table-layout: fixed;`). TBH i was quite surprised by how little improvement minimalCss showed even originally. i have no other explanation for how bootstrap can be faster other than chrome optimizes some stuff specifically for the widely-used bootstrap :/

**Comment by krausest** [comment]
But - maybe a very different and really minimal CSS would be interesting. I came up with this very impressive looking layout:
<img width="556" alt="bildschirmfoto 2017-07-12 um 21 54 12" src="https://user-images.githubusercontent.com/3374055/28137194-b24042b0-674c-11e7-8db5-f57b210c6fdb.png">
Above all it avoids alternating row colors, which leads to the following result (median values, 10 runs):
<img width="1043" alt="bildschirmfoto 2017-07-12 um 21 55 16" src="https://user-images.githubusercontent.com/3374055/28137270-f67abd84-674c-11e7-8e3d-2dbf2de55705.png">
At least it mixes advantages and disadvantages of keyed vs. non-keyed a bit.The factors are a bit higher than for bootstrap but not as strong as I expected.
I'll commit that CSS as the new minimal version.

**Comment by leeoniya** [comment]
![so-beutiful](https://user-images.githubusercontent.com/43234/28139197-f3a57ad8-6718-11e7-8104-6f1847d27586.jpg)

**Comment by krausest** [comment]
Maybe I should consider this bootstrap theme: https://news.ycombinator.com/item?id=14941043

**Comment by krausest** [comment]
I'm closing some older issues that won't get fixed.
